### PR TITLE
Caching tests

### PR DESCRIPTION
**Pull request by hepcat72**
## Summary Change Description

Added tests to the caching code and fixes a few bugs (that didn't affect regular caching operations).

- Fixed a test in test_models.py
- Removed the cache building methods from HierCachedModel because I couldn't figure out how to do it without importing the derived classes.
- Made get_my_cached_method_names a class method.
- Changed the use_cache variable to caching_retrievals because it was too confusing in the testing.
- Created a delete_all_caches method so that files using this file don't also need to import django.core.cache
- Changed get_result to not consult the root record via caches_exist, because I realized that you cannot infer cache invalidity from the representative - only validity.  Any invalidation is/must be handled via immediate deletion of all affected cached values.

## Affected Issue Numbers

- Partially addresses #336

## Code Review Notes

- I removed efforts to be able to build caches from HierCachedModel because I could not figure out how to do it elegantly.
- Changed delete_cache to a split into 2 different methods: delete_related_caches (which deletes all from the root) and delete_descendant_caches (which deletes from the caller downward)

## Checklist

- [ ] All issue requirements satisfied (or no linked issues)
- [x] [Linting passes](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting).
- [x] [Migrations created & committed *(or no model changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)
- [x] [Tests implemented *(or no code changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
- [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)

### CODE REVIEW

**Comment by hepcat72** [comment]
Oh gee.  More linting failures that don't show up locally...

**Comment by lparsons** [comment]
> Oh gee. More linting failures that don't show up locally...

Very strange. I haven't run into this. Do you pull the latest super-linter every time? It only seems to update every couple of months for me, but it did just recently.

**Comment by hepcat72** [comment]
> > Oh gee. More linting failures that don't show up locally...
> 
> Very strange. I haven't run into this. Do you pull the latest super-linter every time? It only seems to update every couple of months for me, but it did just recently.

I think this is a false alarm.  I re-copied the load commands I'd used in the new files from test_compositeviews.py due to the rebase.  Must have silently changed the indent level where I pasted because black changes it...

**Comment by hepcat72** [comment]
OK, passes now...

**Review by jcmatese — APPROVED** [review]
thanks for the tests. feel free to press onward!

File: DataRepo/tests/test_hier_cached_model.py [hunk 1]
```diff
@@ -0,0 +1,597 @@
...
+        sample_table_headers="DataRepo/example_data/sample_table_headers.yaml",
+    )
+    call_command(
+        "load_accucor_msruns",
+        protocol="Default",
+        accucor_file="DataRepo/example_data/small_dataset/small_obob_maven_6eaas_inf.xlsx",
+        date="2021-06-03",
+        researcher="Michael Neinast",
+        new_researcher=True,
+    )
```
Comment (lparsons) [inline DataRepo/tests/test_hier_cached_model.py:54]: Comment: Not part of this PR, but we really should get the example data organized into "studies" so we can just call `load_study`.

File: DataRepo/tests/test_hier_cached_model.py [hunk 2]
```diff
@@ -0,0 +1,597 @@
...
+    maxDiff = None
+
+    @classmethod
+    def setUpTestData(cls):
+        load_data()
+
+    def test_load_not_cached(self):
+        a = Animal.objects.all().first()
+        f = "final_serum_sample"
+        v, s = get_cache(a, f)
```
Comment (lparsons) [inline DataRepo/tests/test_hier_cached_model.py:67]: NB: Non-breaking, but using single letter variable names makes this hard for me to follow. I understand they are much faster to type, but I find these fairly hard to decipher. I can follow `a` == "animal" and `f` == "final serum sample", but I'm not sure I know what `v` and `s` stand for without reading further in code (value and status).

**Review by lparsons — APPROVED** [review]
This looks like it's fairly comprehensive, and I trust you knew which things were the most likely to be problematic, so I'm good with moving ahead. 👍🏻

File: DataRepo/tests/test_hier_cached_model.py (hunk 2 above)
Comment (hepcat72) [reply DataRepo/tests/test_hier_cached_model.py:67]: Yeah, too much of a shortcut.  For now, v=value, f=function, s=status.

- [ ] Flesh out the variable names

Comment (hepcat72) [reply DataRepo/tests/test_hier_cached_model.py:67]: Imncidentally, I had to change `l` because something complained it was "ambiguous", though how it's ambiguous, I don't know.  Perhaps there's a builtin by that name?

**Review by fkang-pu — APPROVED** [review]
The tests seem to cover all cases, which are also helpful to me to visualize the workflow of caching. Nice work, Rob!

**Comment by hepcat72** [comment]
> The tests seem to cover all cases, which are also helpful to me to visualize the workflow of caching. Nice work, Rob!

Note, I was overly explicit in the tests, so all that code is not strictly necessary.  For example, this test:

```
    def test_get_cache_enabled(self):
         a = Animal.objects.all().first()
         f = "final_serum_sample"
         # Ensure I get the value not using the cache
         disable_caching_retrievals()
         v = getattr(a, f)  # same as `v = a.final_serum_sample`
         delete_all_caches()
         enable_caching_updates()
         set_cache(a, "final_serum_sample", v)
         enable_caching_retrievals()
         res, sts = get_cache(a, "final_serum_sample")
         self.assertEqual(
             res,
             v,
             msg="Cached value should be correct when caching retrievals are enabled and the value is cached",
         )
         self.assertTrue(
             sts,
             msg="Cached status should be true when caching retrievals are enabled and the value is cached",
         )
```

Could be accomplished simply with:
```
    def test_get_cache_enabled(self):
         a = Animal.objects.all().first()
         set_cache(a, "final_serum_sample", v)
         res, sts = get_cache(a, "final_serum_sample")
         self.assertEqual(
             res,
             v,
             msg="Cached value should be correct when caching retrievals are enabled and the value is cached",
         )
         self.assertTrue(
             sts,
             msg="Cached status should be true when caching retrievals are enabled and the value is cached",
         )
```

I just didn't want to make any assumptions in the test, like if someone changed the hard-coded global defaults.

In fact, the test could even assume more and just do:

```
    def test_get_cache_enabled(self):
         res = Animal.objects.all().first().final_serum_sample
         self.assertEqual(
             res,
             "expected value",
             msg="Cached value should be correct when caching retrievals are enabled and the value is cached",
         )
```

...but that assumes that if the caching code wasn't working, it would return a bad value...
